cin_validator/rules/cin2022_23/rule_8873.py

```diff
@@ -0,0 +1,290 @@
+from cin_validator.utils import make_census_period
+
+# Get tables and columns of interest from the CINTable object defined in rule_engine/__api.py
+
+Assessments = CINTable.Assessments
+LAchildID = Assessments.LAchildID
+AssessmentFactors = Assessments.AssessmentFactors
+CINdetailsID = Assessments.CINdetailsID
+AssessmentActualStartDate = Assessments.AssessmentActualStartDate
+
+CINdetails = CINTable.CINdetails
+LAchildID = CINdetails.LAchildID
+CINdetailsID_cin = CINdetails.CINdetailsID
+ReasonForClosure = CINdetails.ReasonForClosure
+
+# Reference date in header is needed to define the period of census.
+
+
+# define characteristics of rule
+@rule_definition(
+    # write the rule code here, in place of 2885
+    code=8873,
+    # replace ChildProtectionPlans with the value in the module column of the excel sheet corresponding to this rule .
+    # Note that even if multiple tables are involved, one table will be named in the module column.
+    module=CINTable.ChildProtectionPlans,
+    # replace the message with the corresponding value for this rule, gotten from the excel sheet.
+    message="When there is only one assessment on the episode and the factors code “21 No factors identified” has been used for the completed assessment, the reason for closure ‘RC8’ must be used.",
+    # The column names tend to be the words within the < > signs in the github issue description.
+    affected_fields=[
+        ReasonForClosure,
+    ],
+)
+def validate(
+    data_container: Mapping[CINTable, pd.DataFrame], rule_context: RuleContext
+):
+    # PREPARING DATA
+
+    # Replace ChildProtectionPlans with the name of the table you need.
+    df_ass = data_container[Assessments].copy()
+    df_cin = data_container[CINdetails].copy()
+
+    # Before you begin, rename the index so that the initial row positions can be kept intact.
+    df_ass.index.name = "ROW_ID"
+    df_cin.index.name = "ROW_ID"
+
+    # Resetting the index causes the ROW_IDs to become columns of their respective DataFrames
+    # so that they can come along when the merge is done.
+    # TODO summarise with a for loop? e.g for df in [df_cpp, df_47, df_cin]
+    df_ass.reset_index(inplace=True)
+    df_cin.reset_index(inplace=True)
+
+    # lOGIC
+    # Within a <CINDetails> group, if there is only one <Assessment> group present and <AssessmentFactors> (N00181) = “21”, <ReasonForClosure> (N00103) must = RC8.
+
+    # Eliminates rows with more than 1 assessment per CINdetails group by determining if there's more than 1 AssessmentActualStartDate per CINdetailsID per child
+    df_ass_merged = df_ass.merge(df_ass, on=["LAchildID", "CINdetailsID"])
+    df_ass_merged = df_ass_merged[
+        (
+            df_ass_merged["AssessmentActualStartDate_x"]
+            != df_ass_merged["AssessmentActualStartDate_y"]
+        )
+    ]
+    more_than_1_ass = df_ass_merged["ROW_ID_x"].tolist()
+
+    df_ass = df_ass[~df_ass["ROW_ID"].isin(more_than_1_ass)]
+
```

Comment:
This works well. Another way you could have done it is to groupby the IDs and create a column of the count of Assessments per group.  Then filter based on the rows whose count is more than one. like this:

use transform() in the groupby to ensure that the series created in the result has the same dimension as other columns in the dataframe so that it fits in nicely. That is, the count value will be expanded such that the same count value is repeated for each row in a CIN group.
`df_ass["num_ass"] = df_ass.groupby(["LAchildID", "CINdetailsID"])["AssessmentActualStartDate"].transform("count")`

filter
`df_ass = df_ass[df_ass["num_ass"]<2]`

Now we have a df_ass with only the groups where 1 assessment occurs.